revisions-information.md

```diff
@@ -0,0 +1,74 @@
+
+Revisions can also be used when a package homepage or maintainer or the like changes, to keep information displayed current.
+
+## Why not just keep uploading new versions?
+
+First, uploading a whole bunch of code when only metadata changes leads to an unnecessary growth in versions. Second, often revisions need to be applied not only to the most recent version of a package, but prior versions as well. In particular, if a package at a given version has a bad install plan, then you do not want to let some tool continue to think this is a good plan, even if that package is not the latest version,.
+
+## Where can I see revisions?
+
+Packages with revisions will tell you so on their cabal page. You can always access the revisions of a package by appending `/revisions` to its url, like so: http://hackage.haskell.org/package/lens-3.5.1/revisions/
+
+## How can I download a package without revisions applied?
+
+`cabal get --pristine`
+
+## How are revisions numbered?
+
+Revisions are numbered by natural numbers, starting at zero, and increase in increments of 1.
+
+## What can revisions change?
+
+Revisions are intended to only update metadata used by build tooling, or that is used purely for display purposes. The source of a package is never changed, only specified fields of cabal files. Given the same fixed dependencies and flag configuration, as long as a metadata change hasn't disallowed them, the build product from a compiler should be the same regardless of revision changes.
+
+Specifically, revisions can alter the following:
+
+* Dependency bounds (tightening or loosening)
+* Cabal spec version (only between the interval 1.10 and 1.21)
+* Custom-setup dependencies (adding, removing or altering bounds)
+* Flag defaults
+* Flag type (automatic to manual only)
+* Flag descriptions
+* Dependencies (adding, but only in very special cases, see below).
+* Copyright attribution
+* Maintainer attribution
+* Author attribution
+* Tested-with information
+* Bug report URL
+* Homepage URL
+* Source repositories
+* Package synopsis
+* Package description
+* Package category
+
+Revisions can also add custom-setup stanzas. This is necessary because newer cabal library versions make explicit custom-setup depends, while older versions simply shared dependencies with the package proper.
+
+Other exceptions for adding dependencies all fall in the camp of needing to add things which had been implicitly assumed in the past, from a small whitelist. Package dependencies which may be added are "base" and "base-orphans". Build tool dependencies which may be added are "alex", "c2hs", "cpphs", "greencard", "happy" and "hsc2hs".
```

Comment:
This is similarly become more relevant with newer cabals which require accurate dependency tracking, since cabal has been shifting to a less "stateful" package environment; what happened to work accidentally in the past, now doesn't anymore; and with the advent of qualified goals (c.f. https://www.well-typed.com/blog/2015/03/qualified-goals/) we now solve build tools independently, and this also allows each package in an install plan to have the exact version of `alex` it requires; but this comes at the cost of requiring more accurate `build-tools`/`build-tool-depends` depedencies; all we're doing here is making implicit assumptions more explicit; we're not adding something "new".